source/posix/thread.c

```diff
@@ -336,22 +375,22 @@ int aws_thread_launch(
         pthread_attr_destroy(attributes_ptr);
     }
 
-    if (attr_return == EINVAL) {
-        return aws_raise_error(AWS_ERROR_THREAD_INVALID_SETTINGS);
-    }
-
-    if (attr_return == EAGAIN) {
-        return aws_raise_error(AWS_ERROR_THREAD_INSUFFICIENT_RESOURCE);
-    }
-
-    if (attr_return == EPERM) {
-        return aws_raise_error(AWS_ERROR_THREAD_NO_PERMISSIONS);
-    }
-
-    if (allocation_failed || attr_return == ENOMEM) {
-        return aws_raise_error(AWS_ERROR_OOM);
+    if (attr_return) {
+        s_thread_wrapper_destroy(wrapper);
+
+        switch (attr_return) {
+            case EINVAL:
+                return aws_raise_error(AWS_ERROR_THREAD_INVALID_SETTINGS);
+            case EAGAIN:
+                return aws_raise_error(AWS_ERROR_THREAD_INSUFFICIENT_RESOURCE);
+            case EPERM:
+                return aws_raise_error(AWS_ERROR_THREAD_NO_PERMISSIONS);
+            case ENOMEM:
+                return aws_raise_error(AWS_ERROR_OOM);
+            default:
+                return aws_raise_error(AWS_ERROR_UNKNOWN);
+        }
     }
-
     return AWS_OP_SUCCESS;
```

Comment:
Fix some fishy error handling.
The previous code forgot to free memory, and could report success if the error wasn't one of the 4 it expected